### PR TITLE
fixed #10 by trying url patterns

### LINKED ISSUE #10: AUR4 的下载 URL 改变了
见 https://github.com/archlinuxcn/repo/issues/110#issuecomment-110287270

### PR DESCRIPTION
**Pull request by farseerfc**
To solve #10 we will try 3 url patterns now. We may need to re-order the templates list after migration to aur4 is done.

### CODE REVIEW

**Comment by lilydjwg** [comment]
There's no need to HEAD first. Just get its content and check if it's 200. Then write the data to tar (or maybe use `tarfile`?).

**Comment by farseerfc** [comment]
Changed to write data through tarfile. This will save some external command call and file operations.
I added a AurDownloadException which will be raised instead of CalledProcessException from `tar` command, but I am not sure where to catch this exception. 
I added handling for it in lilac, is it enough?

File: lilaclib.py [hunk 1]
```diff
@@ -111,20 +117,41 @@ def download_official_pkgbuild(name):
...
+  aur3url = 'https://aur.archlinux.org/packages/{first_two}/{name}/{name}.tar.gz'
+  aur4url = 'https://aur.archlinux.org/cgit/aur.git/snapshot/{name}.tar.gz'
+  templates = [aur4mig, aur3url, aur4url]  # TODO: reorder this after migration to aur4
+  urls = [url.format(first_two=name[:2], name=name) for url in templates]
+  for url in urls:
+    response = requests.get(url)
+    if response.status_code == 200:
+      logger.debug("downloaded aur tarball '%s' from url '%s'" % (name, url))
+      return response.content
+  logger.error("failed to find aur url for '%s'" % name)
```
Comment (lilydjwg) [inline lilaclib.py]: 你不需要自己格式化字符串，这个工作是 logging 内部会在需要的时候做的。

File: lilaclib.py [hunk 2]
```diff
@@ -111,20 +117,41 @@ def download_official_pkgbuild(name):
...
+  files = []
+  try:
+    tar_file = tarfile.open(name=name+".tar.gz", mode="r:gz", fileobj=tar_content)
+    for filename in tar_file.getnames():
+      basename, remain = os.path.split(filename)
+      if basename == '':
+        continue
+      if remain in ('.AURINFO', '.SRCINFO'):
+        continue
+      contentfile = open(remain, "wb")
```
Comment (lilydjwg) [inline lilaclib.py]: use `with`.

File: lilaclib.py [hunk 3]
```diff
@@ -111,20 +117,41 @@ def download_official_pkgbuild(name):
...
+      contentfile.write(tar_file.extractfile(filename).read())
+      contentfile.close()
+      files.append(remain)
+  finally:
+    if tar_file:
+      tar_file.close()
+    if tar_content:
+      tar_content.close()
```
Comment (lilydjwg) [inline lilaclib.py]: No need to close a `BytesIO`. Also tarfiles support `with` so you don't need to try.

**Pull request merged**